### ISSUE TITLE
Regression in jaxlib 0.7.2, CPU and GPU: donated buffer no longer updated in place when a loop reads the intermediate value, peak memory grows by a full array copy

### ISSUE THREAD

**Issue by Gattocrucco** [openxla/xla#45912]
(Written by Claude Fable 5, lightly edited by me.)

## Description

Since jaxlib 0.7.2, XLA allocates an extra temp buffer the size of a donated input for the pattern

```
donated x  →  elementwise update A  →  while loop reads A (read-only)  →  elementwise update B  →  output (aliased to x)
```

Up to jaxlib 0.7.1 the whole chain `x_initial → x_intermediate → x_final` was computed in place in the donated allocation and `temp_size_in_bytes` was ~0. From 0.7.2 the fusion computing `x_intermediate` is assigned a fresh temp allocation of the full array size, so peak memory is ~2x the working set. This happens on both CPU and GPU, caused by two different commits that expose the same underlying buffer-assignment limitation. Still present in the latest release (tested jax 0.11.0, CPU and GPU/CUDA 12).

## Reproducer

```python
from functools import partial
import jax
import jax.numpy as jnp
from jax import lax

N = 1 << 27  # f32 -> 512 MiB

@partial(jax.jit, donate_argnums=0)
def f(x_initial):
    x_intermediate = jnp.exp(x_initial)  # exp is expensive, so it is not rematerialized into the final fusion
    
    def body(i, s):
        return s + x_intermediate[i * (N // 8)]
    s = lax.fori_loop(0, 8, body, 0.0)   # keeps `x_intermediate` live across the loop

    m = jnp.full((16, 16), s)
    for _ in range(4):
        m = m @ m                        # small-gemm padding, see note below

    x_final = x_intermediate + s         # output 0 is aliased into x_initial
    return x_final, m

x = jax.ShapeDtypeStruct((N,), jnp.float32)
m = f.lower(x).compile().memory_analysis()
print(jax.__version__, m.temp_size_in_bytes)
```

`temp_size_in_bytes` (4·N = 536870912 is one full copy of `x_initial`):

| jax/jaxlib | CPU | GPU |
|---|---|---|
| 0.7.1 | 1 088 | 2 560 |
| 0.7.1.dev20250820 (nightly) | 1 088 | — |
| 0.7.2.dev20250821 (nightly) | **536 870 980** | 2 560 |
| 0.7.2.dev20250822 (nightly) | — | **536 871 172** |
| 0.7.2 | 536 870 980 | 536 871 172 |
| 0.7.2 with `--xla_gpu_experimental_enable_command_buffer_on_thunks=false` | — | 2 560 |
| latest (0.11.0) | 536 870 980 | 536 871 172 |

(The gemm chain is padding so that, on ≤ 0.7.1, the post-loop region reaches `xla_gpu_graph_min_graph_size` and the final fusion lands in a command buffer; without it the GPU shows the extra copy on all versions. It is inert on CPU.)

## Bisection

Mixing jax/jaxlib wheels across each nightly boundary gives good behavior with new jax + old jaxlib and bad with old jax + new jaxlib, so both regressions are in the XLA compiler, not JAX lowering. The two culprits:

- **CPU** (nightly flip 20250820 → 20250821): ade8738371ca69fe5ade3193147b2644edd76b95 *"[xla:cpu] Do not outline HLO instruction with parallel config into separate computations"*.

  Confirmed structurally: the good build outlines the large fusions into `parallel_*` computations invoked by `kCall`; the bad build keeps them inline with `outer_dimension_partitions` set directly.

- **GPU** (nightly flip 20250821 → 20250822): 9decd2af5882d6fe73f2f91ca59d14e0da58d41b *"Enable creation of command buffers on thunks by default."*, which disables the HLO-level `CommandBufferScheduling` pass.

  Confirmed causally: flipping `xla_gpu_experimental_enable_command_buffer_on_thunks` back to false on the bad build restores in-place reuse (that escape hatch disappeared when the pass was removed in 174a8d291b6abe5a0f74afa3e19c3dbfe8295254, Dec 2025). Structurally: the good build's HLO has `%call = (...) call(%wrapped_exponential, %while...), to_apply=%command_buffer`; the bad build has a flat entry computation.

## Analysis

Both commits are reasonable on their own terms (the thunk runtimes don't need HLO-level outlining). What they both removed, incidentally, is a `kCall` wrapping around the consuming fusion — and that wrapping was what let buffer assignment prove the in-place chain safe.

There is no `copy` instruction in any of the four builds; the difference is in `BufferAssigner`. The live ranges of `x_intermediate` and `x_final` abut exactly (`x_intermediate.end == x_final.start`, the `x_final` fusion is `x_intermediate`'s last use), so sharing the donated allocation hinges on the `can_share_as_operand` check in `BufferAssigner::LiveRangeInterferes` (backend-independent):

https://github.com/openxla/xla/blob/3ede8358116e1795f21e4a8833920ae5ce9f8fdc/xla/service/buffer_assignment.cc#L1831-L1845

which is applied when the two live ranges abut exactly:

https://github.com/openxla/xla/blob/3ede8358116e1795f21e4a8833920ae5ce9f8fdc/xla/service/buffer_assignment.cc#L1847-L1857

`operand_end_position` is the last `HloPosition` of `x_intermediate`'s value, computed by `HloLiveRange::ComputeValueLiveRangeEnd`:

https://github.com/openxla/xla/blob/3ede8358116e1795f21e4a8833920ae5ce9f8fdc/xla/hlo/utils/hlo_live_range.cc#L156-L186

Because the while loop carries `x_intermediate` through its tuple (and reads it through `bitcast` reshapes), the value's last position is a while-related position (or a bitcast), which the `x_final` fusion is **not** a direct user of, so `IsUserOf(...)` fails and the abutting ranges are declared interfering — even though the `x_final` fusion is a loop fusion whose use of `x_intermediate` is elementwise, i.e. `HloDataflowAnalysis::CanShareOperandBufferWithUser` on the (defining instruction, fusion) pair itself would return true:

https://github.com/openxla/xla/blob/3ede8358116e1795f21e4a8833920ae5ce9f8fdc/xla/hlo/analysis/hlo_dataflow_analysis.cc#L1684-L1692

(elementwise loop fusions are accepted by the fallback at the end of that function:)

https://github.com/openxla/xla/blob/3ede8358116e1795f21e4a8833920ae5ce9f8fdc/xla/hlo/analysis/hlo_dataflow_analysis.cc#L1853-L1855

With the old outlining (parallel-task calls on CPU, command buffers on GPU), dataflow propagated `x_intermediate`'s value into the callee parameter of the outlined computation containing the `x_final` fusion; that parameter, scheduled after the loop, became the value's last position, and the fused root *is* a direct user of it — so the check passed and the whole chain was computed in place.

So the underlying issue is that `can_share_as_operand` only considers the value's last *position* instead of the actual user relationship at the abutment point (or all positions). The fact that two independent HLO-to-thunk migrations, on two backends, each regressed donated-buffer reuse through this same check suggests fixing the check itself. A fix there would also make in-place reuse robust for other schedules where a read-only consumer (loop, bitcast view) happens to hold the value's last position.

Possibly related: jax-ml/jax#32028 reports a GPU `temp_size_in_bytes` jump in the same 0.7.0 → 0.7.2 window, appearing above a batch-size threshold — consistent with the command-buffer min-graph-size threshold no longer being reached at the HLO level.

**Comment by akuegel** [comment]
Thank you for the report with the thorough analysis. I think you/the AI is right that this is a shortcoming in buffer assignment. https://github.com/openxla/xla/pull/45947 is the attempt to fix it, still in internal review.

**Comment by akuegel** [comment]
> So the underlying issue is that can_share_as_operand only considers the value's last position instead of the actual user relationship at the abutment point (or all positions). The fact that two independent HLO-to-thunk migrations, on two backends, each regressed donated-buffer reuse through this same check suggests fixing the check itself. A fix there would also make in-place reuse robust for other schedules where a read-only consumer (loop, bitcast view) happens to hold the value's last position.

I think that conclusion is wrong. We need to figure out why the sharing did happen before but not now, but at least this was not the reason.

**Comment by Gattocrucco** [comment]
I asked Claude Opus 4.8 to check if #45947 fixes the problem. Its report, cpu-only:
- the PR does fix the problem in the issue MWE
- however, the unit tests added by the PR do not actually test this regression, they pass even without the fix
- the missing ingredient in the unit tests is something something committed scheduling

While when filing the issue I started from something I found myself and pointed Claude at it, this time I used Opus (I'm out of Fable this week) and I am just reporting its first answer without checking, so take it for what it is.

**Comment by akuegel** [comment]
@Gattocrucco You are right, the test does not test this regression. But I did try a long time to come up with a HLO-level reproducer and failed, even a direct dump of what buffer assignment prints when running the JAX reproducer did not show that the change is needed. In the end it turned out that it depends on the schedule. In the test I need to use sequential ordering so that the instructions are processed in exactly that order that is dumped during the JAX reproducer. Now I got a simplified test that fails without the fix, and passes with the fix. I will merge it soon.

**Comment by akuegel** [comment]
Should be fixed now with https://github.com/openxla/xla/commit/7bb7b1e8a44acaa5c0e9158296fc38c344e3fb3a
